synapse/storage/databases/main/user_directory.py

```diff
@@ -663,6 +663,42 @@ async def get_user_dir_rooms_user_is_in(self, user_id):
         users.update(rows)
         return list(users)
 
+    @cached()
+    async def get_shared_rooms_for_users(self, user_id, other_user_id):
+        """
+        Returns the rooms that a user is in.
+
+        Args:
+            user_id(str): Must be a local user
+
+        Returns:
+            list: user_id
+        """
+        SQL = """
+            SELECT p1.room_id
+            FROM users_in_public_rooms as p1
```

Comment:
I'm afraid there's still the slight niggle of this code relying on the user directory updating processes, in that we should tell users that if it's [disabled](https://github.com/matrix-org/synapse/blob/bd6dc17221741d4ceae05ae769a70696ae939336/synapse/handlers/user_directory.py#L95) then this will break as well.

This [config option](https://github.com/matrix-org/synapse/blob/5c43c43240a85ca6b65ad327b6a5b1c9e29bd653/synapse/config/server.py#L141-L143) controls that variable. It's not listed in the sample config file as it's really only used when you're splitting user directory handling out to a user_dir worker.

Needless to say, we should document this somewhere. The place to start would be to note that `^/_matrix/client/unstable/user/.*/shared_rooms/` is also [handled by the user directory worker](https://github.com/matrix-org/synapse/blob/develop/docs/workers.md#synapseappuser_dir).